pkg/ingester/ingester.go

```diff
@@ -468,6 +468,45 @@ func (i *Ingester) starting(ctx context.Context) (err error) {
 		return errors.Wrap(err, "opening existing TSDBs")
 	}
 
+	if i.ownedSeriesService != nil {
+		// We need to perform the initial computation of owned series after the TSDBs are opened but before the ingester becomes
+		// ACTIVE in the ring and starts to accept requests. However, because the ingester still uses the Lifecycler (rather
+		// than BasicLifecycler) there is no deterministic way to delay the ACTIVE state until we finish the calculations.
+		//
+		// Since we don't actually need to be ACTIVE in the ring to calculate owned series (just present, with tokens) we instead
+		// calculate owned series before we start the lifecycler at all. Once we move the ingester to the BasicLifecycler and
+		// have better control over the ring states, we could perform the calculation while in JOINING state, and move to
+		// ACTIVE once we finish.
+
+		// TODO: move logic to owned_series.go?
+
+		oss := i.ownedSeriesService
+
+		// Fetch and cache current ring state
+		_, err := oss.checkRingForChanges()
```

Comment:
Why not do this in starting method of owned series service?

Reply:
The issue we're trying to prevent making worse is the ingester becoming `ACTIVE` in the ring before the ingester service is `Running`. This issue exists even without the owned series service, but the potentially lengthy calculation of owned series being done after the lifecycler is started (which starts the clock on the ingester becoming ACTIVE) but before the service switches to `Running` makes it more likely that we start accepting requests and rejecting them because the ingester service is still `Starting`.

Doing the owned series calculation before starting the lifecycler ensures that the calculation doesn't increase the delay between those two state changes.

Reply:
I understand what you're asying @pr00se but, from a design perspective, this logic should still be in `oss.starting()`. What you can simply do here, is to start `oss` and wait until running, while moving this code block back to `oss.starting()`.

Reply:
@pracucci I think it's a bit more complicated than that. Because [the `services.Manager` requires all subservices to be in `New` state](https://github.com/grafana/dskit/blob/b9a439dedeb88a6ce8f90f6d7cb37e7fdd3330ab/services/manager.go#L325-L327), starting `oss` here would mean we can no longer use the manager to watch and handle its state. Since all we really needed was to run an iteration of the `oss` loop before starting it, I opted to pull the `starting` logic out rather than change the whole service management aspect, though I agree that it's not the cleanest design.

If you feel strongly I can move this back to `starting` and handle the state of the `oss` outside of the manager. I assume I can base the code off of how we handle the `Lifecycler` already.

Reply:
This was done in https://github.com/grafana/mimir/pull/7583